### ISSUE TITLE
Unstable order when reading spec from `.json` file

### ISSUE THREAD

**Issue by PhilipTrauner** [CreateAPI/CreateAPI#12]
https://github.com/kean/CreateAPI/blob/87ba65d28a657ef7b94a758697cbf68d6075dbb9/Sources/CreateAPI/Generate.swift#L159-L160

In addition to not being thread-safe, `JSONDecoder` also does not appear to guarantee any explicit order of object keys.

<details>
<summary>Spec</summary>

```json
{
  "openapi": "3.0.0",
  "info": {
    "version": "1.0.0",
    "title": "Order"
  },
  "paths": {
    "/container": {
      "get": {
        "responses": {
          "200": {
            "description": "",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Container"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "Container": {
        "type": "object",
        "required": null,
        "properties": {
          "propertyOne": {
            "type": "string"
          },
          "propertyTwo": {
            "type": "string"
          },
          "propertyThree": {
            "type": "string"
          },
          "propertyFour": {
            "type": "string"
          }
        }
      }
    }
  }
}
```
</details>

<details>
<summary>Output</summary>

```swift
// Generated by Create API
// https://github.com/kean/CreateAPI
//
// swiftlint:disable all

import Foundation

public struct Container: Codable {
    public var propertyFour: String?
    public var propertyThree: String?
    public var propertyOne: String?
    public var propertyTwo: String?

    public init(propertyFour: String? = nil, propertyThree: String? = nil, propertyOne: String? = nil, propertyTwo: String? = nil) {
        self.propertyFour = propertyFour
        self.propertyThree = propertyThree
        self.propertyOne = propertyOne
        self.propertyTwo = propertyTwo
    }
}

struct StringCodingKey: CodingKey, ExpressibleByStringLiteral {
    private let string: String
    private var int: Int?

    var stringValue: String { return string }

    init(string: String) {
        self.string = string
    }

    init?(stringValue: String) {
        self.string = stringValue
    }

    var intValue: Int? { return int }

    init?(intValue: Int) {
        self.string = String(describing: intValue)
        self.int = intValue
    }

    init(stringLiteral value: String) {
        self.string = value
    }
}
```
</details>

The easiest workaround is to convert the `.json` spec to a `.yaml` spec beforehand.

**Comment by kean** [comment]
Unfortunately, that seems to be the case. It's recommended to use YAML. [OpenAPIKit](https://github.com/mattpolzin/OpenAPIKit) uses ordered dictionaries to preserve the order of the keys when it can (YAML).

There is also an option to sort properties alphabetically which is disabled by default:

```yaml
entities:
  isSortingPropertiesAlphabetically: true
```

**Comment by PhilipTrauner** [comment]
> The primary objective of this revision is to bring YAML into compliance with JSON as an official subset.
> 
> — [YAML 1.2 Specification](https://yaml.org/spec/1.2.0/) (2009)

It *should* be perfectly valid to use `YAMLDecoder` for `.json` documents. 
This of course hinges on how closely [Yams](https://github.com/jpsim/Yams) follows the spec.

I didn't run into any issues with the OpenAPI specs I tried (two small internal ones), but that feels like too small of a sample size to conclusively rule out any nasty edge cases.

**Comment by PhilipTrauner** [comment]
> If retaining order is important for your use-case, I recommend the [Yams](https://github.com/jpsim/Yams) and [FineJSON](https://github.com/omochi/FineJSON) libraries for YAML and JSON respectively. Also keep in mind that JSON is entirely valid YAML and therefore you will likely get good results from Yams decoding of JSON as well (it just won't encode as valid JSON).
>
> — [OpenAPIKit Documentation](https://github.com/mattpolzin/OpenAPIKit#a-note-on-dictionary-ordering)

Not sure if "you will likely get good results" is good enough, but I'd argue that entity generation with JSON as input should currently be considered broken, as its behavior is unexpected for anyone unfamiliar with the JSON specification.

Would you be interested in a PR that switches the JSON parser over to `Yams` and adds tests? 
It wouldn't be necessary to check in any JSON specs, as decoding a YAML spec with `Yams` and consequently encoding it with `JSONEncoder` does retain order.

**Comment by liamnichols** [comment]
I had a go at doing this, but I found a bug in Yams that is a _slight_ blocker (it caused one of the specs that decodes a json configuration to fail):

- https://github.com/jpsim/Yams/pull/351

Other than the empty string/optional issue though, Yams has been pretty good at handling our 31,500 line OpenAPI schema in JSON format, we've just been renaming it to .yml but not converting it from JSON.

**Comment by pjlogan39** [comment]
> https://github.com/kean/CreateAPI/blob/87ba65d28a657ef7b94a758697cbf68d6075dbb9/Sources/CreateAPI/Generate.swift#L159-L160
> 
> In addition to not being thread-safe, `JSONDecoder` also does not appear to guarantee any explicit order of object keys.
> 
> Spec
> Output
> The easiest workaround is to convert the `.json` spec to a `.yaml` spec beforehand.

Nigger

**Comment by liamnichols** [comment]
Fixed in [0.0.4](https://github.com/CreateAPI/CreateAPI/releases/tag/0.0.4)! Thanks 🚀
